std/algorithm/comparison.d

```diff
@@ -1565,3 +1568,129 @@ unittest
                 2, "two",
                 ));
 }
+
+/**
+    Returns true if all of the elements in $(D r1) are present in $(D r2).
+    Otherwise, returns false.
+
+    If the passed ranges are random access ranges, then this function will
+    not allocate GC handled memory. But, if your parameters are forward ranges,
+    then this function can allocate if the $(D AllocateGC.yes) enum is passed.
+    This has the benefit of being much faster than the $(D AllocateGC.no) option.
+
+    Params:
+        r1 = Can be any finite forward or random access range
+        r2 = Can be any finite forward or random access range
+
+    Returns:
+        $(D bool)
+*/
+enum AllocateGC
+{
+    yes,
+    no
+};
+
+bool isPermutation(Range1, Range2)(Range1 r1, Range2 r2, AllocateGC allocate_gc = AllocateGC.no) if (
+    isForwardRange!Range1 &&
+    isForwardRange!Range2 &&
+    !isRandomAccessRange!Range1 &&
+    !isRandomAccessRange!Range2 &&
+    !isInfinite!Range1 &&
+    !isInfinite!Range1)
+{
+    if (allocate_gc == AllocateGC.yes)
+    {
+        import std.conv: to;
+
+        alias E1 = Unqual!(ElementType!Range1);
+        alias E2 = Unqual!(ElementType!Range2);
+
+        // If the two ranges aren't the same length, then they cannot
+        // possibly be permutations
+        static if (hasMember!(Range1, "length") && hasMember!(Range2, "length"))
+        {
+            if (r1.length != r2.length) {
+                return false;
+            }
+        }
+
+        int[E1] r1_counts;
+        int[E2] r2_counts;
+
+        foreach (item; r1)
+        {
+            ++r1_counts[item];
+        }
+
+        foreach (item; r2)
+        {
+            ++r2_counts[item];
+        }
+
+        return r1_counts == r2_counts;
+    }
+    else
+    {
+        import std.algorithm.searching : count;
+
+        static if (hasMember!(Range1, "length") && hasMember!(Range2, "length"))
+        {
+            if (r1.length != r2.length) {
+                return false;
+            }
+        }
+
+        foreach (item; r1.save)
+        {
+            if (r1.save.count(item) != r2.save.count(item))
+            {
+                return false;
+            }
+        }
```

Comment:
Hmm, just realized that the above two examples are wrong, we should iterate over a `save`d copy of `r1`, not `r1` itself, otherwise the counts will be off.